File: app/src/androidTest/java/com/github/jameshnsears/quoteunquote/WidgetDeletedTest.java
package com.github.jameshnsears.quoteunquote;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.github.jameshnsears.quoteunquote.utils.ContentSelection;
import com.github.jameshnsears.quoteunquote.utils.widget.WidgetIdHelper;

import org.junit.Test;
import org.junit.runner.RunWith;

import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.spy;


@RunWith(AndroidJUnit4.class)
public class WidgetDeletedTest extends QuoteUnquoteModelUtility {
    @Test
    public void twoWidgetsOneDeleted() {

        insertDataset01();
        insertDataset02();

        quoteUnquoteModelDouble.setDefault(WidgetIdHelper.INSTANCE_01_WIDGET_ID);

        quoteUnquoteModelDouble.toggleFavourite(
                WidgetIdHelper.INSTANCE_01_WIDGET_ID,
                quoteUnquoteModelDouble.getNext(
                        WidgetIdHelper.INSTANCE_01_WIDGET_ID, ContentSelection.ALL).digest);

        quoteUnquoteModelDouble.reportQuotation(WidgetIdHelper.INSTANCE_01_WIDGET_ID);

        quoteUnquoteModelDouble.setDefault(WidgetIdHelper.INSTANCE_02_WIDGET_ID);

        final QuoteUnquoteModelDouble quoteUnquoteModelSpy = spy(quoteUnquoteModelDouble);
        doReturn(false).when(quoteUnquoteModelSpy).selectedContentTypeIsFavourite(eq(WidgetIdHelper.INSTANCE_02_WIDGET_ID));

        quoteUnquoteModelSpy.toggleFavourite(
                WidgetIdHelper.INSTANCE_02_WIDGET_ID,
                quoteUnquoteModelDouble.getNext(
                        WidgetIdHelper.INSTANCE_01_WIDGET_ID, ContentSelection.ALL).digest);

        quoteUnquoteModelDouble.reportQuotation(WidgetIdHelper.INSTANCE_02_WIDGET_ID);

        assertEquals(
                "",
                1,
                quoteUnquoteModelDouble.countPrevious(
                        WidgetIdHelper.INSTANCE_01_WIDGET_ID, ContentSelection.ALL));

        assertEquals(
                "",
                1,
                quoteUnquoteModelDouble.countPrevious(
                        WidgetIdHelper.INSTANCE_02_WIDGET_ID, ContentSelection.ALL));

        assertEquals(
                "",
                0,
                quoteUnquoteModelDouble.countFavourites());

        assertEquals(
                "",
                1,
                quoteUnquoteModelDouble.countReported());

        quoteUnquoteModelSpy.toggleFavourite(
                WidgetIdHelper.INSTANCE_02_WIDGET_ID,
                quoteUnquoteModelDouble.getNext(WidgetIdHelper.INSTANCE_01_WIDGET_ID, ContentSelection.ALL).digest);
        assertEquals(
                "",
                1,
                quoteUnquoteModelDouble.countFavourites());

        ////////////////////////////////////////////

        quoteUnquoteModelDouble.delete(WidgetIdHelper.INSTANCE_01_WIDGET_ID);

        assertEquals(
                "",
                0,
                quoteUnquoteModelDouble.countPrevious(WidgetIdHelper.INSTANCE_01_WIDGET_ID, ContentSelection.ALL));

        assertEquals(
                "",
                1,
                quoteUnquoteModelDouble.countFavourites());

        assertEquals(
                "",
                1,
                quoteUnquoteModelDouble.countReported());

        ////////////////////////////////////////////

        quoteUnquoteModelDouble.disable();

        assertEquals(
                "",
                0,
                quoteUnquoteModelDouble.countPrevious(WidgetIdHelper.INSTANCE_02_WIDGET_ID, ContentSelection.ALL));

        assertEquals(
                "",
                0,
                quoteUnquoteModelDouble.countFavourites());

        assertEquals(
                "",
                0,
                quoteUnquoteModelDouble.countReported());
    }
}

File: app/src/androidTest/java/com/github/jameshnsears/quoteunquote/WidgetDeletedTest.kt
package com.github.jameshnsears.quoteunquote

import com.github.jameshnsears.quoteunquote.configure.fragment.appearance.AppearancePreferences
import com.github.jameshnsears.quoteunquote.configure.fragment.sync.SyncPreferences
import com.github.jameshnsears.quoteunquote.utils.preference.PreferencesFacade
import com.github.jameshnsears.quoteunquote.utils.widget.WidgetIdHelper
import io.mockk.every
import io.mockk.spyk
import junit.framework.TestCase.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

class WidgetDeletedTest : QuoteUnquoteModelUtility() {
    @Test
    fun widgetDeleted() {
        if (canWorkWithMockk()) {
            // arrange
            setupDatabase()
            setupSharedPreferences()

            // act
            val quoteUnquoteWidget = spyk<QuoteUnquoteWidget>()
            every {
                quoteUnquoteWidget.getQuoteUnquoteModel(WidgetIdHelper.WIDGET_ID_01, any())
            } returns quoteUnquoteModelDouble

            quoteUnquoteWidget.onEnabled(context)
            val syncPreferences =
                SyncPreferences(
                    0,
                    context,
                )
            assertTrue(syncPreferences.transferLocalCode.length == 10)
            quoteUnquoteWidget.onDeleted(context, intArrayOf(WidgetIdHelper.WIDGET_ID_01))

            // assert
            assertTrue(quoteUnquoteModelDouble.countPrevious(WidgetIdHelper.WIDGET_ID_01) == 0)
            assertTrue(quoteUnquoteModelDouble.countPrevious(WidgetIdHelper.WIDGET_ID_02) == 3)
            assertTrue(quoteUnquoteModelDouble.countFavourites().blockingGet() == 1)

            assertEquals(
                0,
                PreferencesFacade.countPreferences(context, WidgetIdHelper.WIDGET_ID_01),
            )
            assertEquals(
                1,
                PreferencesFacade.countPreferences(context, WidgetIdHelper.WIDGET_ID_02),
            )
            assertFalse(syncPreferences.transferLocalCode.isEmpty())
        }
    }

    private fun setupDatabase() {
        insertQuotationTestData01()

        setDefaultQuotationAll(WidgetIdHelper.WIDGET_ID_01)
        setDefaultQuotationAuthor(WidgetIdHelper.WIDGET_ID_01)
        setDefaultQuotationSearch(WidgetIdHelper.WIDGET_ID_01)
        assertTrue(quoteUnquoteModelDouble.countPrevious(WidgetIdHelper.WIDGET_ID_01) == 3)

        setDefaultQuotationAll(WidgetIdHelper.WIDGET_ID_02)
        setDefaultQuotationAuthor(WidgetIdHelper.WIDGET_ID_02)
        setDefaultQuotationSearch(WidgetIdHelper.WIDGET_ID_02)
        assertTrue(quoteUnquoteModelDouble.countPrevious(WidgetIdHelper.WIDGET_ID_02) == 3)

        markDefaultQuotationAsFavourite()
        assertTrue(quoteUnquoteModelDouble.countFavourites().blockingGet() == 1)

        // double check only 1 still a favourite
        markDefaultQuotationAsFavourite()
        assertTrue(quoteUnquoteModelDouble.countFavourites().blockingGet() == 1)
    }

    private fun setupSharedPreferences() {
        AppearancePreferences(WidgetIdHelper.WIDGET_ID_01, context).appearanceQuotationTextSize = 1
        AppearancePreferences(WidgetIdHelper.WIDGET_ID_02, context).appearanceQuotationTextSize = 1
    }
}
